include/rmm/mr/device/detail/arena.hpp

```diff
@@ -457,62 +497,50 @@ class arena {
    * @param bytes The size in bytes of the allocation. This must be equal to the value of `bytes`
    * that was passed to the `allocate` call that returned `p`.
    * @param stream Stream on which to perform deallocation.
-   * @return true if the allocation is found, false otherwise.
    */
-  bool deallocate(void* p, std::size_t bytes, cuda_stream_view stream)
+  void deallocate(void* p, std::size_t bytes, cuda_stream_view stream)
   {
     lock_guard lock(mtx_);
-#ifdef RMM_POOL_TRACK_ALLOCATIONS
-    auto const b = free_block(p, bytes);
-#else
     block const b{p, bytes};
-#endif
-    if (b.is_valid()) {
-      auto const merged = coalesce_block(free_blocks_, b);
-      shrink_arena(merged, stream);
-    }
-    return b.is_valid();
+    auto const merged = coalesce_block(free_blocks_, b);
+    shrink_arena(merged, stream);
   }
 
-#ifdef RMM_POOL_TRACK_ALLOCATIONS
   /**
-   * @brief Deallocate memory pointed to by `p`, keeping all free superblocks.
-   *
-   * This is done when deallocating from another arena. Since we don't have access to the CUDA
-   * stream associated with this arena, we don't coalesce the freed block and return it directly to
-   * the global arena.
-   *
-   * @param p Pointer to be deallocated.
-   * @param bytes The size in bytes of the allocation. This must be equal to the value of `bytes`
-   * that was passed to the `allocate` call that returned `p`.
-   * @return true if the allocation is found, false otherwise.
+   * @brief Clean the arena and deallocate free blocks from the global arena.
    */
-  bool deallocate(void* p, std::size_t bytes)
+  void clean()
   {
     lock_guard lock(mtx_);
-    auto const b = free_block(p, bytes);
-    if (b.is_valid()) { global_arena_.deallocate(b); }
-    return b.is_valid();
+    global_arena_.deallocate(free_blocks_);
+    free_blocks_.clear();
   }
-#endif
 
   /**
-   * @brief Clean the arena and deallocate free blocks from the global arena.
+   * Dump memory to log.
    *
-   * This is only needed when a per-thread arena is about to die.
+   * @param logger the spdlog logger to use
    */
-  void clean()
+  void dump_memory_log(std::shared_ptr<spdlog::logger> const& logger) const
   {
     lock_guard lock(mtx_);
-    global_arena_.deallocate(free_blocks_);
-    free_blocks_.clear();
-#ifdef RMM_POOL_TRACK_ALLOCATIONS
-    allocated_blocks_.clear();
-#endif
+    logger->info("    # free blocks: {}", free_blocks_.size());
+    if (!free_blocks_.empty()) {
+      std::size_t total_free{};
+      for (auto const& b : free_blocks_) {
+        total_free += b.size();
+      }
```

Comment:
Same as above. This is the third time this code is repeated. Consider creating a 
```cpp
auto total_free(std::set<block> const& blocks) -> std::size_t;
``` 
helper function.

Reply:
Done.